wgpu-hal/src/gles/egl.rs

```diff
@@ -758,8 +764,47 @@ impl crate::Instance<super::Api> for Instance {
         let mut inner = self.inner.lock();
 
         match raw_window_handle {
-            Rwh::Xlib(_) => {}
-            Rwh::Xcb(_) => {}
+            #[cfg(not(feature = "emscripten"))]
+            Rwh::Xlib(handle) => {
+                // If Wayland EGLDisplay was created, but we got an X window, recreate the context
+                // https://github.com/gfx-rs/wgpu/issues/2762
+                if inner.wl_display.is_some() || inner.wsi.kind == WindowKind::Wayland {
+                    log::warn!("Re-initializing Gles context due to Xlib window");
```

Comment:
One interesting case to consider is that wgpu takes a RawWindowHandle. Nothing actually prevents you from creating a surface using x11 and another with Wayland in the same application.

Vulkan handles that properly but not egl.

Reply:
Maybe a solution could be to add the following to create_surface:

> All created surfaces must have the same type of RawWindowHandle

Reply:
Hmm, exactly! Should it be just a comment or a warning emitted on second change of context type?

Reply:
I guess there is the need to create something like The Book of Wgpu to document such edge cases and the overall design of an application.